### PR TITLE
TRA-4122 - OAS dialog 2

### PR DESCRIPTION
**Pull request by IgorGov**

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (lirazyehezkel) [inline package-lock.json:1]: delete file

File: ui/package.json [hunk 2]
```diff
@@ -31,6 +33,8 @@
     "react-scrollable-feed-virtualized": "^1.4.9",
     "react-syntax-highlighter": "^15.4.3",
     "react-toastify": "^8.0.3",
+    "redoc": "^2.0.0-rc.59",
```
Comment (lirazyehezkel) [inline ui/package.json:36]: try to change version for making snyk checks pass

File: ui/src/components/OasModal/OasModal.tsx [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+const OasModal = ({ openModal, handleCloseModal, entries }) => {
+  const [oasServices, setOASservices] = useState([]);
+  const [selectedOASService, setSelectedOASService] = useState("");
+  const [serviceOAS, setServiceOAS] = useState(null);
+
+  useEffect(() => {
+    (async () => {
+      try {
+        const services = await api.getOASAServices();
+        if (!areEqual(oasServices, services)) setOASservices(services);
```
Comment (lirazyehezkel) [inline ui/src/components/OasModal/OasModal.tsx:19]: check if you a function like this in lodash library

File: ui/src/components/OasModal/OasModal.tsx [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  const onSelectedOASService = async (selectedService) => {
+    setSelectedOASService(selectedService);
+    if(oasServices.length === 0){
+      return
+    }
+    try {
+      const data = await api.getOASAByService(selectedService);
+      setServiceOAS(data);
+    } catch (e) {
+      console.error(e);
```
Comment (lirazyehezkel) [inline ui/src/components/OasModal/OasModal.tsx:35]: display a toast with error

File: ui/src/components/OasModal/OasModal.tsx [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+                    </MenuItem>
+                  ))}
+                </Select>
+              </FormControl>
+            </div>
+            <div style={{ cursor: "pointer" }}>
+              <img src={closeIcon} alt="Back" onClick={handleCloseModal} />
+            </div>
+          </div>
+          {serviceOAS && <RedocStandalone spec={serviceOAS} />}
```
Comment (lirazyehezkel) [inline ui/src/components/OasModal/OasModal.tsx:92]: think about displaying a placeholder text whenever you don't have a serviceOAS selected yet

File: ui/src/components/style/TrafficPage.sass [hunk 6]
```diff
@@ -13,7 +13,12 @@
         display: flex
         align-items: center
         background-color: $header-background-color
+        justify-content: space-between
 
+        .TrafficPageStreameStatus
+                display: flex;
```
Comment (lirazyehezkel) [inline ui/src/components/style/TrafficPage.sass:19]: remove ;

File: ui/src/components/style/TrafficPage.sass [hunk 7]
```diff
@@ -13,7 +13,12 @@
...
+                align-items: center;
```
Comment (lirazyehezkel) [inline ui/src/components/style/TrafficPage.sass:20]: remove ;

File: ui/src/components/TrafficPage.tsx [hunk 8]
```diff
@@ -18,36 +18,35 @@ import entriesAtom from "../recoil/entries";
...
+  setAnalyzeStatus?: (status: any) => void;
+  onTLSDetected: (destAddress: string) => void;
 }
 
-const api = Api.getInstance();
-
-export const TrafficPage: React.FC<TrafficPageProps> = ({onTLSDetected, setAnalyzeStatus}) => {
-
+const api = new Api();
+export const TrafficPage: React.FC<TrafficPageProps> = ({setAnalyzeStatus,onTLSDetected,}) => {
```
Comment (lirazyehezkel) [inline ui/src/components/TrafficPage.tsx:49]: remove the last ','

File: ui/src/components/TrafficPage.tsx [hunk 9]
```diff
@@ -18,36 +18,35 @@ import entriesAtom from "../recoil/entries";
...
-    setAnalyzeStatus?: (status: any) => void;
+  setAnalyzeStatus?: (status: any) => void;
+  onTLSDetected: (destAddress: string) => void;
 }
 
-const api = Api.getInstance();
-
-export const TrafficPage: React.FC<TrafficPageProps> = ({onTLSDetected, setAnalyzeStatus}) => {
-
+const api = new Api();
```
Comment (lirazyehezkel) [inline ui/src/components/TrafficPage.tsx:48]: change to:
const api = Api.getInstance();

File: ui/src/components/TrafficPage.tsx [hunk 10]
```diff
@@ -235,17 +241,41 @@ export const TrafficPage: React.FC<TrafficPageProps> = ({onTLSDetected, setAnaly
...
-                    src={pauseIcon} onClick={toggleConnection}/>
-                <img className="playPauseIcon" style={{position: "absolute", visibility: wsConnection === WsConnectionStatus.Connected ? "hidden" : "visible"}} alt="play"
-                    src={playIcon} onClick={toggleConnection}/>
-                <div className="connectionText">
-                    {getConnectionTitle()}
-                    <div className={"indicatorContainer " + getConnectionStatusClass(true)}>
-                        <div className={"indicator " + getConnectionStatusClass(false)}/>
-                    </div>
-                </div>
+              <div className="TrafficPageStreameStatus">
```
Comment (lirazyehezkel) [inline ui/src/components/TrafficPage.tsx:244]: streame -> stream

File: ui/src/components/TrafficPage.tsx [hunk 11]
```diff
@@ -235,17 +241,41 @@ export const TrafficPage: React.FC<TrafficPageProps> = ({onTLSDetected, setAnaly
...
+                  <img className="playPauseIcon" style={{position: "absolute", visibility: wsConnection === WsConnectionStatus.Connected ? "hidden" : "visible"}} alt="play"
+                      src={playIcon} onClick={toggleConnection}/>
+                  <div className="connectionText">
+                      {getConnectionTitle()}
+                      <div className={"indicatorContainer " + getConnectionStatusClass(true)}>
+                          <div className={"indicator " + getConnectionStatusClass(false)}/>
+                      </div>
+                  </div>
+              </div>
+              {!window["isOasEnabled"] && <div>
```
Comment (lirazyehezkel) [inline ui/src/components/TrafficPage.tsx:256]: remove the !

File: ui/src/components/TrafficPage.tsx [hunk 12]
```diff
@@ -235,17 +241,41 @@ export const TrafficPage: React.FC<TrafficPageProps> = ({onTLSDetected, setAnaly
...
+                    color: "#fff",
+                    textTransform: "none",
+                  }}
+                  onClick={handleOpenModal}
+                >
+                  Show OAS
+                </Button>
+              </div>}
             </div>
+            {!window["isOasEnabled"] && <OasModal
```
Comment (lirazyehezkel) [inline ui/src/components/TrafficPage.tsx:274]: remove the !

File: ui/src/helpers/api.js [hunk 13]
```diff
@@ -61,6 +61,21 @@ export default class Api {
         return response.data;
     }
 
+    getOASAServices = async () => {
```
Comment (lirazyehezkel) [inline ui/src/helpers/api.js:64]: do you use in the end all of those endpoints?

File: ui/src/components/OasModal/OasModal.sass [hunk 14]
```diff
@@ -0,0 +1,6 @@
+@import '../../variables.module.scss'
+
+.NotSelectedMessage
+    margin-left: 41%;
```
Comment (lirazyehezkel) [inline ui/src/components/OasModal/OasModal.sass:4]: sass files can't include the ';' char

File: ui/src/components/OasModal/OasModal.tsx [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+  const onSelectedOASService = async (selectedService) => {
+    setSelectedOASService(selectedService);
+    if(oasServices.length === 0){
+      return
+    }
+    try {
+      const data = await api.getOASAByService(selectedService);
+      setServiceOAS(data);
+    } catch (e) {
+      console.error(e);
```
Comment (lirazyehezkel) [inline ui/src/components/OasModal/OasModal.tsx:40]: you marked this as resolved by I still don't see the toast

File: ui/src/helpers/utils.js [hunk 16]
```diff
@@ -0,0 +1,8 @@
+import intersection from 'lodash/intersection';
+
+export function arrayElementsComparetion(array1,array2){
```
Comment (lirazyehezkel) [inline ui/src/helpers/utils.js:3]: change to ES6 syntax :
export const arrayElementsComparetion = (array1,array2) => {

**Review by lirazyehezkel — CHANGES_REQUESTED** [review]
added some more small comments

File: ui/src/components/OasModal/OasModal.tsx [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+import { toast } from 'react-toastify';
+
+const api = new Api();
+
+const OasModal = ({ openModal, handleCloseModal, entries }) => {
+  const [oasServices, setOASservices] = useState([]);
+  const [selectedOASService, setSelectedOASService] = useState("");
+  const [oasService, setServiceOAS] = useState(null);
+
+  const noOasServiceSelectedMessage = "Please Select OasService";
```
Comment (lirazyehezkel) [inline ui/src/components/OasModal/OasModal.tsx:17]: you can move you to above the component

File: ui/src/helpers/api.js [hunk 18]
```diff
@@ -61,6 +61,16 @@ export default class Api {
...
+        const response = await this.client.get("/oas");
+        return response.data;
+    }
+
+    getOASAByService = async (selectedService) => {
```
Comment (lirazyehezkel) [inline ui/src/helpers/api.js:69]: what is the A after the OAS?

File: ui/src/components/TrafficPage.tsx [hunk 19]
```diff
@@ -235,17 +242,41 @@ export const TrafficPage: React.FC<TrafficPageProps> = ({onTLSDetected, setAnaly
...
+                  onClick={handleOpenModal}
+                >
+                  Show OAS
+                </Button>
+              </div>}
             </div>
+            {window["isOasEnabled"] && <OasModal
+              openModal={openOasModal}
+              handleCloseModal={handleCloseModal}
+              entries={entries}
```
Comment (lirazyehezkel) [inline ui/src/components/TrafficPage.tsx:278]: you don't need to pass entries - you can take it from recoil at OasModal component

File: ui/src/components/OasModal/OasModal.tsx [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+    (async () => {
+      try {
+        const services = await api.getOASAServices();
+        if (!arrayElementsComparetion(oasServices, services)) setOASservices(services);
+      } catch (e) {
+        toast.error(e.message);
+        console.error(e);
+      }
+    })();
+  }, [entries,oasServices]);
```
Comment (lirazyehezkel) [inline ui/src/components/OasModal/OasModal.tsx:29]: please re-think about the logic here:
1. it looks like you fetch the services on each new entry - that means a LOT of requests all the time
2. you can't put oasServices as dep where you use setOasServices

File: ui/src/components/OasModal/OasModal.tsx (hunk 17 above)
Comment (lirazyehezkel) [reply ui/src/components/OasModal/OasModal.tsx:17]: actually I think you can put this text directly in line 87

File: ui/src/components/OasModal/OasModal.tsx (hunk 20 above)
Comment (AmitUp9) [reply ui/src/components/OasModal/OasModal.tsx:29]: 1. I know its a lot of requests, the thing is, that I want to update the oasServices when there is new service. The only way that a new service is arrived is through new entries isn't it? maybe I can call the useEffect every time the "openModal" changes. It means that while someone in the modal the services wont updated and he will need to close and open again. what do you think? 
2. Its make sense but its weird, when I remove it I get the next warning in the console: 
"src/components/OasModal/OasModal.tsx
  Line 28:6:  React Hook useEffect has a missing dependency: 'oasServices'. Either include it or remove the dependency array  react-hooks/exhaustive-deps"

Comment (lirazyehezkel) [reply ui/src/components/OasModal/OasModal.tsx:29]: 1. that sounds like a good solution - go for it

Comment (lirazyehezkel) [reply ui/src/components/OasModal/OasModal.tsx:29]: 2. think about useCallback with the response of api.getOASAServices() as a dependency. that means you will have a useEffect that set a state holds the response of this API and you will update oasServices only via the useCallback

Comment (AmitUp9) [reply ui/src/components/OasModal/OasModal.tsx:29]: What do you think about 2.?

Comment (AmitUp9) [reply ui/src/components/OasModal/OasModal.tsx:29]: did you meant like this? 
  
const getOasServices = useCallback(async () => {
    try {
      const services = await api.getOasServices();
      if (!arrayElementsComparetion(oasServices, services)) setOASservices(services);
    } catch (e) {
      toast.error(e.message);
      console.error(e);
    }
  },[]);

  useEffect(() => {
      getOasServices();
  },[openModal]);

Comment (lirazyehezkel) [reply ui/src/components/OasModal/OasModal.tsx:29]: I thought about putting this const services = await api.getOasServices();
inside the useEffect and then put the service response in a state and then pass this state as a dep for the useCallback.
but the solution you wrote may work as well - test it and if all looks good, we can go with ur solution

File: ui/src/components/OasModal/OasModal.tsx [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+  const onSelectedOASService = async (selectedService) => {
+    setSelectedServiceName(selectedService);
+    if(oasServices.length === 0){
+      return
+    }
+    try {
+      const data = await api.getOasByService(selectedService);
+      setSelectedServiceSpec(data);
+    } catch (e) {
+      toast.error(e.message);
```
Comment (lirazyehezkel) [inline ui/src/components/OasModal/OasModal.tsx:39]: this is a customer-facing message - replace the e.message with a readable message
for example: toast.error("Error occurred while fetching service OAS spec");

File: ui/src/components/OasModal/OasModal.tsx [hunk 22]
```diff
@@ -0,0 +1,95 @@
...
+  const [selectedServiceName, setSelectedServiceName] = useState("");
+  const [selectedServiceSpec, setSelectedServiceSpec] = useState(null);
+
+  useEffect(() => {
+    (async () => {
+      try {
+        const services = await api.getOasServices();
+        setOasServices(services);
+      } catch (e) {
+        toast.error(e.message);
```
Comment (lirazyehezkel) [inline ui/src/components/OasModal/OasModal.tsx:24]: this is a customer-facing message - replace the e.message with a readable message
for example: toast.error("Error occurred while fetching services list");

**Review by lirazyehezkel — APPROVED** [review]

**Pull request merged**